02-intermediate-ruby/inheritance-idioms.md

```diff
@@ -0,0 +1,255 @@
+
+    mortgage = (asking_price - downpayment) * ((1.0 + interest_rate) ** loan_years)
+
+    annual_price = mortgage / loan_years
+    annual_price += annual_property_tax
+
+    return annual_price / 12
+  end
+end
+```
+
+Next, the subclasses. Since the algorithm for calculating payments is already defined in `Property`, we can build these without worrying about that. Instead, we fill in only the details needed for this specific type of property.
+
+```ruby
+# condo.rb
+require_relative 'property'
+
+class Condo < Property
+  # the default value for minimum_downpayment is fine,
+  # so we don't override it here
+
+  # However, we have to implement annual_property_tax
+  def annual_property_tax
+    return asking_price * 0.1
+  end
+end
+```
+
+```ruby
+# mansion.rb
+require_relative 'property'
+
+class Mansion < Property
+  # A mansion requires a big downpayment!
+  def minimum_downpayment
+    return 0.3
+  end
+
+  # The tax calculation is complex
+  def annual_property_tax
+    if asking_price > 1e7
+      return asking_price * 0.3
+    else
+      return asking_price * 0.2
+    end
+  end
+end
+```
+
+Finally, some driver code. Three things to note:
+- We don't instantiate `Property`, only its subclasses
+  - Because of this, we don't need to `require_relative 'property'`
+- We don't call the template methods ourselves, instead relying on `monthly_payment` to know when to do so
+- When we treat `Condo`s and `Mansion`s in the same way, we are taking advantage of _polymorphism_
+
+```ruby
+# main.rb
+
+require_relative 'Condo'
+require_relative 'Mansion'
+
+properties = [
+  Condo.new("123 2nd ave W #206, Seattle, WA", 825_000),
+  Condo.new("307 E Madison #614, Seattle, WA", 614_000),
+  Mansion.new("504 Brickman blvd, Seattle, WA", 1_650_000)
+]
+
+puts "We have many marvelous properties avaiable!"
+
+puts "All figures assume a the minimum downpayment on a 30-year loan with a 4% interest rate"
+
+properties.each do |property|
+  # We don't care if the property is an apartment or a condo
+  puts
+  puts property.address
+  payment = property.monthly_payment(30, 0.04)
+  puts "Estimated monthly payment: $#{payment.round(0)}"
+end
+```
+
+## Summary
+
+Vocabulary
+
+ Term            | Definition                                                 
+-----------------|------------------------------------------------------------
+ software idiom  | A common pattern used by programmers. Not always obvious from the syntax of the language. 
+ abstract class  | A class that is not instantiated, but instead serves as a parent class. 
+ concrete class  | A class that is instantiated. May or may not inherit from some other class. 
+ template method | A method that is intended to be overridden in a subclass.  
+ interface       | Some set of methods that go together. A class that has all those methods _implements_ that interface. 
+ polymorphism    | Treating instances of different classes in the same way. All those classes must implement the same interface.
```

Comment:
I like this page. I'm getting a little greedy here for adding more details. 

Could we add more information on:
- *runtime polymorphism*, which is what is described in the example here.
- *compile time polymorphism*, available in C++, C# and Java
- *interfaces*: There is a recurring theme of confusion around abstract class versus interfaces in Java, especially in the first two weeks of internship.
- *multiple inheritance*: Some languages support it, some don't. Java doesn't and encourages thinking more about the public interfaces the class should implement.

Reply:
I like the idea, but I think it's too much for this point in the curriculum (week 4).